crates/wasmtime/src/runtime/ref/gc_ref/rooting.rs

```diff
@@ -0,0 +1,1423 @@
+            // (Inlined copy of `self.unroot_gc_ref(gc_ref)` to avoid borrowing
+            // `self` while `self.manually_rooted` is already borrowed).
+            //
+            // Safety: our mutable access to the root set means that no one else
+            // should have concurrent access to the `VMExternRef`, so
+            // decrementing the reference count here is safe.
+            unsafe {
+                let _ = extern_ref_from_gc_ref(gc_ref);
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct LifoRoot {
+    generation: u32,
+    gc_ref: VMGcRef,
+}
+
+impl RootSet {
+    /// Enter a LIFO rooting scope.
+    ///
+    /// Returns an integer that should be passed unmodified to `exit_lifo_scope`
+    /// when the scope is finished.
+    ///
+    /// Calls to `{enter,exit}_lifo_scope` must happen in a strict LIFO order.
+    pub(crate) fn enter_lifo_scope(&self) -> usize {
+        let len = self.lifo_roots.len();
+        log::debug!("Entering GC root set LIFO scope: {len}");
+        len
+    }
+
+    /// Exit a LIFO rooting scope.
+    ///
+    /// The `scope` argument must be the result of the corresponding
+    /// `enter_lifo_scope` call.
+    ///
+    /// Calls to `{enter,exit}_lifo_scope` must happen in a strict LIFO order.
+    pub(crate) fn exit_lifo_scope(&mut self, scope: usize) {
+        log::debug!("Exiting GC root set LIFO scope: {scope}");
+        debug_assert!(self.lifo_roots.len() >= scope);
+
+        // In the case where we have a tracing GC, this should really be:
+        //
+        //     self.lifo_roots.truncate(scope);
+        //
+        // In the meantime, without deferred reference-counting collector for
+        // `externref`, we need to drop these references. In the future this
+        // will be a `wasmtime_runtime::GcRuntime::on_unroot` trait method hook
+        // or something like that, but where we can skip this whole loop and
+        // just do the above truncate when the collector doesn't need it.
+        for root in self.lifo_roots.drain(scope..) {
+            // (Inlined copy of `self.unroot_gc_ref(root.gc_ref)` to avoid
+            // borrowing `self` while `self.manually_rooted` is already
+            // borrowed).
+            //
+            // Safety: our mutable access to the root set means that no one else
+            // should have concurrent access to the `VMExternRef`, so
+            // decrementing the reference count here is safe.
+            unsafe {
+                let _ = extern_ref_from_gc_ref(root.gc_ref);
+            }
+        }
+
+        self.lifo_generation += 1;
+    }
+
+    pub(crate) fn with_lifo_scope<T>(
+        store: &mut StoreOpaque,
+        f: impl FnOnce(&mut StoreOpaque) -> T,
+    ) -> T {
+        let scope = store.gc_roots().enter_lifo_scope();
+        let ret = f(store);
+        store.gc_roots_mut().exit_lifo_scope(scope);
+        ret
+    }
+
+    fn unroot_gc_ref(&mut self, gc_ref: VMGcRef) {
```

Comment:
I think this needs to be `unsafe` since there's no guarantee `gc_ref` is valid or belongs to this store?